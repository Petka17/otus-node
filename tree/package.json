{
  "name": "tree",
  "version": "1.0.0",
  "description": "",
  "author": "Petr Klimenko <petrklim@gmail.com>",
  "license": "MIT",
  "type": "module",
  "main": "index.js",
  "bin": {
    "tree": "./src/tree.js"
  },
  "scripts": {
    "test": "jest",
    "test:cov": "jest --coverage"
  },
  "devDependencies": {
    "@babel/core": "^7.10.4",
    "@babel/preset-env": "^7.10.4",
    "babel-jest": "^26.1.0",
    "jest": "^26.1.0",
    "prettier": "^2.0.5"
  }
}
