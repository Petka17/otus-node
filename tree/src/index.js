export const stringifyTree = (tree) =>
  [`${tree.name}`].concat(stringifyTreeItems(tree.items)).join('\n')

const stringifyTreeItems = (treeItems, levels = []) =>
  treeItems.reduce(
    (rows, item, index) => [
      ...rows,
      ...stringifyTreeItem(item, levels.concat(index + 1 === treeItems.length ? 0 : 1)),
    ],
    [],
  )

const stringifyTreeItem = (treeItem, levels) => [
  `${getLevelLines(levels)}-${treeItem.name}`,
  ...(treeItem.items ? stringifyTreeItems(treeItem.items, levels) : []),
]

const getLevelLines = (levels) =>
  levels
    .slice(0, -1)
    .map((level) => (level === 1 ? '| ' : '  '))
    .concat(levels.slice(-1).map((last) => (last === 1 ? '├' : '└')))
    .join('')
