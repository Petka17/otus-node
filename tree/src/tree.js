#!/usr/bin/env node

import path from 'path'

import { constructFolderTree } from './constructFolderTree.js'
import { stringifyTree } from './stringifyTree.js'

const [, , ...argv] = process.argv

console.trace(argv, process.argv)
if (['-h', '--help'].includes(argv[0]) || (argv.length === 2 && argv[0] !== '-d')) {
  console.log('usage: tree [-d folder]')
  process.exit(1)
}

const fullPath = path.resolve(argv[1] ?? '.')

constructFolderTree(fullPath)
  .then(stringifyTree)
  .then(console.log)
  .catch(() => {
    console.error(error)
    process.exit(1)
  })
