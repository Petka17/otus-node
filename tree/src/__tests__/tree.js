jest.mock('path', () => {
  return {
    resolve: () => '/Users/user1/root',
  }
})

jest.mock('../constructFolderTree.js', () => {
  const testData = require('../__test-data__/testDirTree.js')

  return {
    constructFolderTree: () => Promise.resolve(testData),
  }
})

const flushPromises = () => new Promise(setImmediate)

test('returns a tree', async (done) => {
  const consoleLogSpy = jest.spyOn(console, 'log')

  require('../tree')

  await flushPromises()

  expect(consoleLogSpy.mock.calls[0][0]).toMatchSnapshot()

  consoleLogSpy.mockRestore()
  done()
})

// test('show help if -h argument is present', async (done) => {
//   process.argv.push('-h')
//   process.exit = jest.fn(() => {})

//   console.log(process.argv)

//   const consoleLogSpy = jest.spyOn(console, 'log')

//   require('../tree')

//   await flushPromises()

//   expect(process.exit).toHaveBeenCalledWith(1)
//   expect(consoleLogSpy.mock.calls).toBe(1)

//   consoleLogSpy.mockRestore()
//   provess.exit.mockRestore()

//   done()
// })
