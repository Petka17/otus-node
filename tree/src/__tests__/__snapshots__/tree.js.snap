// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a tree 1`] = `
"root
├──child1
│  └──child11
├──child2
└──child3"
`;
